src/session_server.c

```diff
@@ -3080,3 +3098,758 @@ nc_session_get_notif_status(const struct nc_session *session)
+            r = ly_time_time2str(curr_cert->expiration_time, NULL, &exp_time);
+            if (r) {
+                break;
+            }
+
+            /* call the callback */
+            targ->clb(exp_time, curr_cert->xpath, targ->clb_data);
+            free(exp_time);
+
+            /* update the next notification time */
+            curr_cert->notif_time = nc_server_notif_cert_exp_next_notif_time_get(intervals, interval_count, curr_cert);
+            if (curr_cert->notif_time == -1) {
+                break;
+            }
+        } else {
+            ERR(NULL, "Pthread condition timedwait failed (%s).", strerror(r));
+            break;
+        }
+    }
+
+cleanup:
+    VRB(NULL, "Certificate expiration notification thread exit.");
+    if (targ->clb_free_data) {
+        targ->clb_free_data(targ->clb_data);
+    }
+    nc_server_notif_cert_exp_dates_destroy(exp_dates, exp_date_count);
+    free(targ);
+    return NULL;
+}
+
+API int
+nc_server_notif_cert_expiration_thread_start(nc_cert_exp_notif_clb cert_exp_notif_clb,
+        void *user_data, void (*free_data)(void *))
+{
+    int r, ret = 0;
+    pthread_t tid;
+    struct nc_cert_exp_notif_thread_arg *arg;
+
+    NC_CHECK_ARG_RET(NULL, cert_exp_notif_clb, 1);
+
+    /* set the user callback and its data */
+    arg = malloc(sizeof *arg);
+    NC_CHECK_ERRMEM_RET(!arg, 1);
+    arg->clb = cert_exp_notif_clb;
+    arg->clb_data = user_data;
+    arg->clb_free_data = free_data;
+
+    /* LOCK */
+    pthread_mutex_lock(&server_opts.cert_exp_notif.lock);
+
+    /* check if the thread is already running */
+    if (server_opts.cert_exp_notif.thread_running) {
+        ERR(NULL, "Certificate expiration notification thread is already running.");
+        ret = 1;
+        goto cleanup;
+    } else {
+        server_opts.cert_exp_notif.thread_running = 1;
+    }
+
+    if ((r = pthread_create(&tid, NULL, nc_server_notif_cert_exp_thread, arg))) {
+        ERR(NULL, "Creating the certificate expiration notification thread failed (%s).", strerror(r));
+        ret = 1;
+        goto cleanup;
+    }
+
+    server_opts.cert_exp_notif.tid = tid;
+
+cleanup:
+    /* UNLOCK */
+    pthread_mutex_unlock(&server_opts.cert_exp_notif.lock);
+    if (ret) {
+        free(arg);
+    }
+    return ret;
+}
+
+API void
+nc_server_notif_cert_expiration_thread_stop(int wait)
+{
+    int r;
+    pthread_t tid;
+
+    /* LOCK */
+    pthread_mutex_lock(&server_opts.cert_exp_notif.lock);
+    tid = server_opts.cert_exp_notif.tid;
+
+    if (server_opts.cert_exp_notif.thread_running) {
+        /* set the tid and running flag to 0, signal the thread and unlock its mutex */
+        server_opts.cert_exp_notif.thread_running = 0;
+        server_opts.cert_exp_notif.tid = 0;
+        pthread_cond_signal(&server_opts.cert_exp_notif.cond);
+
+        /* UNLOCK */
+        pthread_mutex_unlock(&server_opts.cert_exp_notif.lock);
+        if (wait) {
+            r = pthread_join(tid, NULL);
```

Comment:
To avoid memory leaks, if `!wait`, you need to detach the thread. I am also not sure if not just always wait until the thread finishes, but I have no strong opinion.

Reply:
Added the missing detach, but kept the wait parameter.